aws-java-sdk-sts/src/main/java/com/amazonaws/auth/STSSessionCredentialsProvider.java

```diff
@@ -119,36 +160,119 @@ public STSSessionCredentialsProvider(AWSCredentialsProvider longLivedCredentials
      */
     public void setSTSClientEndpoint(String endpoint) {
         securityTokenService.setEndpoint(endpoint);
-        sessionCredentials = null;
+        sessionCredentials.set(null);
     }
 
     @Override
     public AWSSessionCredentials getCredentials() {
         if (needsNewSession()) startSession();
 
-        return sessionCredentials;
+        return sessionCredentials.get().sessionCredentials;
     }
 
     @Override
     public void refresh() {
-        startSession();
+        asyncNewSession();
```

Comment:
The interface contract is ambiguous in my opinion. To me, it is not strongly-worded and suggests the method is implementation-specific. But your point is valid, I can make it a blocking call.
/**
     * Forces this credentials provider to refresh its credentials. For many
     * implementations of credentials provider, this method may simply be a
     * no-op, such as any credentials provider implementation that vends
     * static/non-changing credentials. For other implementations that vend
     * different credentials through out their lifetime, this method should
     * force the credentials provider to refresh its credentials.
     */